### PR TITLE
feat: refactor ShapeStream API to combine  and  to better support API proxies

### LINKED ISSUE #204: Clients should just ask for the `url` to a shape (not `baseURL` and `table` we currently have)
I've been exploring how apps can query for shapes. I've discovered three patterns so far:

1. Client just asks for the shape and the API proxies it e.g.

```ts
// Client
{
    shape: { table: `contacts` },
    baseUrl: `http://localhost:5173/shape-proxy`,
}

// Server
export async function handler({ params, request }) {
  const requestUrl = new URL(request.url)
  const { table } = params
  
  // Validate user can request this shape
  if (checkShapeAccessPermission(user, shapeDef)) {
    const originUrl = new URL(`http://localhost:3000/v1/shape/${table}`)
    requestUrl.searchParams.forEach((value, key) => {
      originUrl.searchParams.set(key, value)
    })
    return await fetch(originUrl.toString())
  } else {
    // 403 access denied
  }
}
```

2. Client asks for shape and API adds where clauses to restrict access e.g. in this case by the organization the user belongs to.

```ts
// Client
{
    shape: { table: `contacts` },
    baseUrl: `http://localhost:5173/shape-proxy`,
}

// Server
export async function handler({ params, request }) {
  const requestUrl = new URL(request.url)
  const { table } = params
  
  // Validate user can request this shape
  if (checkShapeAccessPermission(user, shapeDef)) {
    const originUrl = new URL(`http://localhost:3000/v1/shape/${table}`)
    requestUrl.searchParams.forEach((value, key) => {
      originUrl.searchParams.set(key, value)
    })

    // Restrict to contacts within the organization the user belongs to.
    originUrl.searchParams.set(`where`, `org_id = ${user.org_id}`)
    return await fetch(originUrl.toString())
  } else {
    // 403 access denied
  }
}
```

3. The third option isn't supported by our shape definition but which is that people could use normal REST API patterns e.g. `/api/contacts`. Then the API would just proxy that to the right URL on electric along with query parameters like `?offset=STRING`. 

In order to support this third pattern, users should just directly specify the backend URL to hit for a shape e.g.

```ts
const shapeDef = {
  url: `/api/shape/contacts`
}
```

This is also more elegant as it combines the `baseURL` and `shape` object into just the `url`.

### PR DESCRIPTION
**Pull request by KyleAMathews**
Fix #204

### CODE REVIEW

File: packages/react-hooks/src/react-hooks.tsx [hunk 1]
```diff
@@ -28,17 +28,14 @@ export async function preloadShape(
...
 
 export function sortedOptionsHash(options: ShapeStreamOptions): string {
-  const shapeDef = JSON.stringify(
-    options.shape,
-    Object.keys(options.shape).sort()
-  )
+  // const shapeDef = JSON.stringify(
+  // options.shape,
+  // Object.keys(options.shape).sort()
+  // )
```
Comment (icehaunter) [inline packages/react-hooks/src/react-hooks.tsx:34]: Forgot to delete or uncomment?

Comment (KyleAMathews) [reply packages/react-hooks/src/react-hooks.tsx]: yeah, fixed

File: docs/api/clients/typescript.md [hunk 2]
```diff
@@ -33,8 +33,7 @@ import { ShapeStream } from '@electric-sql/next'
 
 // Passes subscribers rows as they're inserted, updated, or deleted
 const stream = new ShapeStream({
-  baseUrl: `${BASE_URL}`,
-  shape: { table: `foo` }
+  url: `${BASE_URL}/v1/shape/foo`,
```
Comment (kevin-dp) [inline docs/api/clients/typescript.md:36]: Is it meant to use `BASE_URL` here?
It's not really defined in the example, so perhaps just a localhost url is more clear?

File: packages/react-hooks/README.md [hunk 3]
```diff
@@ -29,8 +31,7 @@ import { useShape } from "@electric-sql/react"
 
 export default function MyComponent () {
   const { isUpToDate, data as fooData } = useShape({
-    shape: { table: `foo` },
-    baseUrl: "http://my-api.com/",
+    url: "http://my-api.com/shape/foo",
```
Comment (kevin-dp) [inline packages/react-hooks/README.md:34]: Shouldn't the url be `"http://my-api.com/v1/shape/foo"` (i added `/v1`)

File: packages/react-hooks/src/react-hooks.tsx [hunk 4]
```diff
@@ -28,17 +28,8 @@ export async function preloadShape(
...
-  )
-  // eslint-disable-next-line
-  const { shape, ...optionsWithoutShapeDef } = options
-  const allOptions = JSON.stringify(
-    optionsWithoutShapeDef,
-    Object.keys(options).sort()
-  )
-  const shapeHash = shapeDef + allOptions
+  const allOptions = JSON.stringify(options, Object.keys(options).sort())
+  const shapeHash = allOptions
```
Comment (kevin-dp) [inline packages/react-hooks/src/react-hooks.tsx:32]: This assignment is not needed, could just return `allOptions`.

Comment (kevin-dp) [reply packages/react-hooks/src/react-hooks.tsx:32]: The trick with providing the keys to be stringified and sorting that array of keys is neat although i'm not sure if it is safe to rely on this trick. I quickly looked at the docs and - unless i missed it - it does not explicitly state that the order of the properties in the stringified output will respect the order of the keys in the reviver argument.

**Review by kevin-dp — APPROVED** [review]
Looks good 👍

File: packages/react-hooks/README.md (hunk 3 above)
Comment (KyleAMathews) [reply packages/react-hooks/README.md:34]: no as the request is being proxied through their API (so the url path is somewhat arbitrary)

File: docs/api/clients/typescript.md (hunk 2 above)
Comment (KyleAMathews) [reply docs/api/clients/typescript.md:36]: yeah true, will fix

File: packages/react-hooks/src/react-hooks.tsx (hunk 4 above)
Comment (KyleAMathews) [reply packages/react-hooks/src/react-hooks.tsx:32]: I was basing it off this stackoverflow answer — https://stackoverflow.com/a/16168003 — it seems pretty likely to me that the stringified output is written in order? It passes the test anyways... unless we have a better approach or you're sure it doesn't work, it seems fine.

Will fix `allOptions`

**Pull request merged**